### ISSUE TITLE
bigtable-hbase-2.x-shaded:2.1.0 : INTERNAL: RST_STREAM closed stream. HTTP/2 error code: INTERNAL_ERROR

### ISSUE THREAD

**Issue by fr-ashikaumagiliya** [googleapis/java-bigtable-hbase#3735]
I use 'bigtable-hbase-2.x-shaded:2.1.0' in my project and I get the following error intermittently :
 
`com.google.bigtable.repackaged.com.google.api.gax.rpc.InternalException: com.google.bigtable.repackaged.io.grpc.StatusRuntimeException: INTERNAL: RST_STREAM closed stream. HTTP/2 error code: INTERNAL_ERROR
`

I noticed a similar bug is fixed here:  https://github.com/googleapis/java-bigtable-hbase/pull/3000
But seems like its for 1.x version. this is fixed in 2.1.0 as well ?

`java.lang.RuntimeException: org.apache.hadoop.hbase.DoNotRetryIOException: Failed to perform operation. Operation='put', projectId='fr-prd-datalake-k8s', tableName='prd_slk_event_tbl', rowKey='00000000000084082478'
	at com.fr.det.datalake.eventdriven.service.EventService.saveToHBase(EventService.java:187)
	at com.fr.det.datalake.eventdriven.service.EventService.saveEvent(EventService.java:134)
	at com.fr.det.datalake.eventdriven.service.EventService$$FastClassBySpringCGLIB$$40f3cc49.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:779)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:388)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:692)
	at com.fr.det.datalake.eventdriven.service.EventService$$EnhancerBySpringCGLIB$$c09fec47.saveEvent(<generated>)
	at com.fr.det.datalake.eventdriven.controller.EventListener.listen(EventListener.java:97)
	at jdk.internal.reflect.GeneratedMethodAccessor82.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:171)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:120)
	at org.springframework.kafka.listener.adapter.HandlerAdapter.invoke(HandlerAdapter.java:48)
	at org.springframework.kafka.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:329)
	at org.springframework.kafka.listener.adapter.RecordMessagingMessageListenerAdapter.onMessage(RecordMessagingMessageListenerAdapter.java:86)
	at org.springframework.kafka.listener.adapter.RecordMessagingMessageListenerAdapter.onMessage(RecordMessagingMessageListenerAdapter.java:51)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeOnMessage(KafkaMessageListenerContainer.java:2065)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeOnMessage(KafkaMessageListenerContainer.java:2047)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeRecordListener(KafkaMessageListenerContainer.java:1984)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeWithRecords(KafkaMessageListenerContainer.java:1924)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeRecordListener(KafkaMessageListenerContainer.java:1812)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeListener(KafkaMessageListenerContainer.java:1531)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.pollAndInvoke(KafkaMessageListenerContainer.java:1178)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.run(KafkaMessageListenerContainer.java:1075)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.apache.hadoop.hbase.DoNotRetryIOException: Failed to perform operation. Operation='put', projectId='fr-prd-datalake-k8s', tableName='prd_slk_event_tbl', rowKey='00000000000084082478'
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.logAndCreateIOException(AbstractBigtableTable.java:563)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.mutateRow(AbstractBigtableTable.java:497)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.put(AbstractBigtableTable.java:357)
	at com.fr.det.datalake.eventdriven.service.EventService.saveToHBase(EventService.java:184)
	... 34 more
Caused by: com.google.bigtable.repackaged.com.google.api.gax.rpc.InternalException: com.google.bigtable.repackaged.io.grpc.StatusRuntimeException: INTERNAL: RST_STREAM closed stream. HTTP/2 error code: INTERNAL_ERROR
	at java.base/java.lang.Thread.getStackTrace(Unknown Source)
	at com.google.cloud.bigtable.hbase.util.FutureUtil.setStackTrace(FutureUtil.java:60)
	at com.google.cloud.bigtable.hbase.util.FutureUtil.rethrow(FutureUtil.java:49)
	at com.google.cloud.bigtable.hbase.util.FutureUtil.unwrap(FutureUtil.java:40)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.mutateRow(AbstractBigtableTable.java:494)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.put(AbstractBigtableTable.java:357)
	at com.fr.det.datalake.eventdriven.service.EventService.saveToHBase(EventService.java:184)
	at com.fr.det.datalake.eventdriven.service.EventService.saveEvent(EventService.java:134)
	at com.fr.det.datalake.eventdriven.service.EventService$$FastClassBySpringCGLIB$$40f3cc49.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:779)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:388)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:692)
	at com.fr.det.datalake.eventdriven.service.EventService$$EnhancerBySpringCGLIB$$c09fec47.saveEvent(<generated>)
	at com.fr.det.datalake.eventdriven.controller.EventListener.listen(EventListener.java:97)
	at jdk.internal.reflect.GeneratedMethodAccessor82.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:171)
	at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:120)
	at org.springframework.kafka.listener.adapter.HandlerAdapter.invoke(HandlerAdapter.java:48)
	at org.springframework.kafka.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:329)
	at org.springframework.kafka.listener.adapter.RecordMessagingMessageListenerAdapter.onMessage(RecordMessagingMessageListenerAdapter.java:86)
	at org.springframework.kafka.listener.adapter.RecordMessagingMessageListenerAdapter.onMessage(RecordMessagingMessageListenerAdapter.java:51)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeOnMessage(KafkaMessageListenerContainer.java:2065)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeOnMessage(KafkaMessageListenerContainer.java:2047)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeRecordListener(KafkaMessageListenerContainer.java:1984)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeWithRecords(KafkaMessageListenerContainer.java:1924)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeRecordListener(KafkaMessageListenerContainer.java:1812)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeListener(KafkaMessageListenerContainer.java:1531)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.pollAndInvoke(KafkaMessageListenerContainer.java:1178)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.run(KafkaMessageListenerContainer.java:1075)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
	at --------Future.get--------(Unknown Source)
	at com.google.bigtable.repackaged.com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:65)
	at com.google.bigtable.repackaged.com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.bigtable.repackaged.com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.bigtable.repackaged.com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.bigtable.repackaged.com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.bigtable.repackaged.com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1133)
	at com.google.bigtable.repackaged.com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
	at com.google.bigtable.repackaged.com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1277)
	at com.google.bigtable.repackaged.com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1038)
	at com.google.bigtable.repackaged.com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:808)
	at com.google.bigtable.repackaged.io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:564)
	at com.google.bigtable.repackaged.io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:534)
	at com.google.bigtable.repackaged.io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at com.google.bigtable.repackaged.io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at com.google.bigtable.repackaged.io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.bigtable.repackaged.com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:455)
	at com.google.bigtable.repackaged.io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at com.google.bigtable.repackaged.io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at com.google.bigtable.repackaged.io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.bigtable.repackaged.io.grpc.census.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:802)
	at com.google.bigtable.repackaged.io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at com.google.bigtable.repackaged.io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at com.google.bigtable.repackaged.io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.bigtable.repackaged.io.grpc.census.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:428)
	at com.google.bigtable.repackaged.io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:562)
	at com.google.bigtable.repackaged.io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at com.google.bigtable.repackaged.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:743)
	at com.google.bigtable.repackaged.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:722)
	at com.google.bigtable.repackaged.io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at com.google.bigtable.repackaged.io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	... 1 more
Caused by: com.google.bigtable.repackaged.io.grpc.StatusRuntimeException: INTERNAL: RST_STREAM closed stream. HTTP/2 error code: INTERNAL_ERROR
	at com.google.bigtable.repackaged.io.grpc.Status.asRuntimeException(Status.java:535)
	... 22 more`

**Comment by ponkon** [comment]
I'm getting a similar error when using `com.google.cloud.bigtable:bigtable-hbase-1.x-hadoop:1.27.1`

```
ERROR c.g.c.b.hbase.AbstractBigtableTable - Encountered exception when executing checkAndPut.com.google.bigtable.repackaged.io.grpc.StatusRuntimeException: INTERNAL: HTTP/2 error code: INTERNAL_ERROR
Received Rst Stream
	at com.google.bigtable.repackaged.io.grpc.Status.asRuntimeException(Status.java:524)
	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.async.AbstractRetryingOperation.getBlockingResult(AbstractRetryingOperation.java:455)
	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.BigtableDataGrpcClient.checkAndMutateRow(BigtableDataGrpcClient.java:291)
	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.BigtableDataClientWrapper.checkAndMutateRow(BigtableDataClientWrapper.java:144)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndMutate(AbstractBigtableTable.java:465)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndPut(AbstractBigtableTable.java:391)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndPut(AbstractBigtableTable.java:370)
	at com.qsc.et.meta.MetaAgent.updateAllMeta(MetaAgent.java:377)
	at com.qsc.et.bt.Loader.loadPackageNative(Loader.java:182)
	at com.qsc.et.bt.Loader.loadPackage(Loader.java:107)
```

**Comment by fr-ashikaumagiliya** [comment]
> I'm getting a similar error when using ~`com.google.cloud.bigtable:bigtable-hbase-1.x-hadoop:1.27.1`~
> 
> Nevermind, was using an older version `com.google.cloud.bigtable:bigtable-hbase-1.x-hadoop:1.14.1`
> 
> ```
> ERROR c.g.c.b.hbase.AbstractBigtableTable - Encountered exception when executing checkAndPut.com.google.bigtable.repackaged.io.grpc.StatusRuntimeException: INTERNAL: HTTP/2 error code: INTERNAL_ERROR
> Received Rst Stream
> 	at com.google.bigtable.repackaged.io.grpc.Status.asRuntimeException(Status.java:524)
> 	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.async.AbstractRetryingOperation.getBlockingResult(AbstractRetryingOperation.java:455)
> 	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.BigtableDataGrpcClient.checkAndMutateRow(BigtableDataGrpcClient.java:291)
> 	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.BigtableDataClientWrapper.checkAndMutateRow(BigtableDataClientWrapper.java:144)
> 	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndMutate(AbstractBigtableTable.java:465)
> 	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndPut(AbstractBigtableTable.java:391)
> 	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndPut(AbstractBigtableTable.java:370)
> 	at com.qsc.et.meta.MetaAgent.updateAllMeta(MetaAgent.java:377)
> 	at com.qsc.et.bt.Loader.loadPackageNative(Loader.java:182)
> 	at com.qsc.et.bt.Loader.loadPackage(Loader.java:107)
> ```

Does it work fine after upgrading ?

**Comment by ponkon** [comment]
Unfortunately no, getting different - but similar errors as before. Some 502s:

```
c.g.c.b.hbase.AbstractBigtableTable - Encountered exception when executing checkAndDelete.com.google.bigtable.repackaged.io.grpc.StatusRuntimeException: UNAVAILABLE: 502:Bad Gateway
	at com.google.bigtable.repackaged.io.grpc.Status.asRuntimeException(Status.java:526)
	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.async.AbstractRetryingOperation.getBlockingResult(AbstractRetryingOperation.java:423)
	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.BigtableDataGrpcClient.checkAndMutateRow(BigtableDataGrpcClient.java:307)
	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.BigtableDataClientWrapper.checkAndMutateRow(BigtableDataClientWrapper.java:144)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndMutate(AbstractBigtableTable.java:465)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndDelete(AbstractBigtableTable.java:436)
```

and some connection reset by peer (which sounds a lot like the Received Rst Stream from before)

```
Caused by: org.apache.hadoop.hbase.DoNotRetryIOException: Failed to perform operation. Operation='checkAndDelete', projectId='**************', tableName='**************', rowKey='****************'
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.logAndCreateIOException(AbstractBigtableTable.java:548)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndMutate(AbstractBigtableTable.java:469)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndDelete(AbstractBigtableTable.java:436)
	at *******************************
	... 19 common frames omitted
Caused by: com.google.bigtable.repackaged.io.grpc.StatusRuntimeException: UNAVAILABLE: io exception
	at com.google.bigtable.repackaged.io.grpc.Status.asRuntimeException(Status.java:526)
	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.async.AbstractRetryingOperation.getBlockingResult(AbstractRetryingOperation.java:423)
	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.BigtableDataGrpcClient.checkAndMutateRow(BigtableDataGrpcClient.java:307)
	at com.google.bigtable.repackaged.com.google.cloud.bigtable.grpc.BigtableDataClientWrapper.checkAndMutateRow(BigtableDataClientWrapper.java:144)
	at com.google.cloud.bigtable.hbase.AbstractBigtableTable.checkAndMutate(AbstractBigtableTable.java:465)
	... 21 common frames omitted
Caused by: com.google.bigtable.repackaged.io.grpc.netty.shaded.io.netty.channel.unix.Errors$NativeIoException: readAddress(..) failed: Connection reset by peer
```

**Comment by mutianf** [comment]
> I noticed a similar bug is fixed here: https://github.com/googleapis/java-bigtable-hbase/pull/3000. But seems like its for 1.x version. this is fixed in 2.1.0 as well ?

The fix in #3000 is also done since version 2.0 but it's only on read rows requests. For mutations, it's fixed in v2.6.2 and v1.27.1.

The fix will convert rst_stream errors to unavailable which is retryable.
